### ISSUE TITLE
Overhaul fontselect: individual, cached, full-detail queries

### ISSUE THREAD

**Issue by astiob** [libass/libass#771]
Extracting from https://github.com/libass/libass/issues/511#issuecomment-873572826:

# Proposed plan

* Stop adding system fonts to the main font list at all.
* Ask the system font provider for fonts every single time a font is needed, instead of only upon failure to find one.
* Pass the full font details (including weight and slant) in each font provider query.
* Have the font provider return a list of fonts, instead of adding them to the main font list.
* Have the font provider cache the list of fonts for each query. Or if this is sufficient, only the one font that the font selector ends up selecting.

# Benefits

This will improve consistency and predictability of font selection across the board:

**We’ll have clear separation between attached fonts and system fonts;** see #509. We could even expose a configuration toggle to enable/disable system font lookup in cases when an attached font exists, so that older broken releases could still be fixed by installing a missing font system-wide.

**We’ll improve fallback font choice** by taking into account the main font’s style rather than just its name.

**We’ll fix corner cases** where we currently load a font into our database that is a bad-but-acceptable match for some later query, where our current logic returns this bad match for this later query without even consulting the system font provider. For example:

1. We want font Foo.
2. We have no font Foo yet.
3. We ask the system font provider for Foo.
4. It finds a font “Foo”, but also happens to find a font “Foo Bar” with an alternative name of “Bar”.
5. Later, we want font Bar.
6. We already have Bar from the earlier lookup, so we don’t ask for more fonts.
7. Actually, the system contains several other Bars and we’ve selected the wrong one.

**We’ll fix more obscure corner cases** where the font provider doesn’t return a font at all if we give it only the name. For example:

1. We have queries for font Foo with several different weights.
2. The system has several Foo fonts installed, but one has Foo as its full name and the others as family name.
3. The system font provider is “DirectWrite with GDI”, so it is only able to find the full-name Foo if we tell it the exact weight it has.
4. We don‘t pass the weight to the font provider, so we never find that font.

# Implementation

Unspecified future. PRs welcome, but this isn’t gonna be trivial.

# Related issues

Font fallback: not only is our fallback logic itself odd as discussed elsewhere, but we cache the resulting fonts in a fallback-priority-list associated with each font named in the ASS, and we continue using the same fallback font for all code points until we find another code point that is entirely missing. But it’s possible that, say, in a line with two foreign-language inserts, the fallback font for the first foreign insert contains glyphs for the second foreign language but they’re terrible, and if we asked the system font provider for a fallback font for the second foreign language, it’d give us some other, better font. I don’t know if this can be solved efficiently, but I’m writing this down anyway.

**Comment by rcombs** [comment]
> Ask the system font provider for fonts every single time a font is needed, instead of only upon failure to find one.

Presumably this excludes cases where an attached font is a name match (and the "allow system fonts to take precedence over muxed ones" option is disabled)?

Also note that this is probably required for Android AFont lookup to work.

**Comment by astiob** [comment]
> Presumably this excludes cases where an attached font is a name match (and the "allow system fonts to take precedence over muxed ones" option is disabled)?

Yes, when/if we want muxed fonts to have priority, we’d perform our lookup on _only_ the muxed fonts first and then query the system font provider if this lookup fails.

When we allow system fonts to override muxed ones, we’d query the system font provider first and then perform our lookup on the combined list of all muxed fonts + whatever the provider returned for this particular query.

**Comment by moi15moi** [comment]
> * Pass the full font details (including weight and slant) in each font provider query.

To be fully compatible with GDI, we may need to parse the OS/2 table (for italic and weight) and the name table (for family and fullname/PostScript) ourselves. It is pretty straightforward on Windows and Mac with [IDWriteFontFace::TryGetFontTable](https://learn.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritefontface-trygetfonttable) and [CTFontCopyTable](https://developer.apple.com/documentation/coretext/1510755-ctfontcopytable?language=objc), but fontconfig doesn't allow querying font tables. The possible future Android provider with the [NDK](https://developer.android.com/ndk/reference/group/font) also doesn't allow querying tables. In those case, I guess we can use freetype to open the entire font and use their api to get the detail we need (like we currently do it for memory font).
This approch may be significally slower then what we currently have (especially for fontconfig and the future android provider).

Finally, this approach adds complexity because we would need to parse font information in two different ways: using the table memory AND with a FreeType face object. Both methods must parse the font identically, resulting in duplicated logic. We need to carefully consider how to handle this properly.

**Comment by MIRIMIRIM** [comment]
Based on my previous tests, the weight obtained from DWrite may not necessarily be the same as the weight in OS/2. I guess that GDI uses the weight from OS/2 (because I haven’t tested GDI)? I would like to know if there are any related issues.

**Comment by astiob** [comment]
I’m not sure I follow, sorry. We already take the fonts, pass them to FreeType and read out the names and style parameters from FreeType ourselves for all fonts except Fontconfig ones.

**Comment by moi15moi** [comment]
> Pass the full font details (including weight and slant) in each font provider query.

I totally misread. I thought it was "parse," not "pass," so I believed the initial goal was to parse the font ourselves and ignore the metadata provided by the provider. Please ignore my message.

> Also, why not switch from DWrite to GDI?

I never found a way to properly implement the method get_fallback with GDI, which is one of the reasons why libass probably uses dwrite. However, yes, all the other features can be implemented. I did it [here](https://github.com/moi15moi/libass/tree/Add-GDI-provider).

The only "good" reason to use GDI instead of dwrite would be to also use the "native Windows font provider" for Windows XP.
